comms/src/protocol/messaging/inbound.rs

```diff
@@ -0,0 +1,118 @@
+    peer_manager::Peer,
+    protocol::messaging::{MessagingEvent, MessagingProtocol},
+};
+use futures::{channel::mpsc, AsyncRead, AsyncWrite, SinkExt, StreamExt};
+use log::*;
+use std::sync::Arc;
+use tokio::sync::broadcast;
+
+const LOG_TARGET: &str = "comms::protocol::messaging::inbound";
+
+pub struct InboundMessaging {
+    peer: Arc<Peer>,
+    inbound_message_tx: mpsc::Sender<InboundMessage>,
+    messaging_events_tx: broadcast::Sender<Arc<MessagingEvent>>,
+}
+
+impl InboundMessaging {
+    pub fn new(
+        peer: Arc<Peer>,
+        inbound_message_tx: mpsc::Sender<InboundMessage>,
+        messaging_events_tx: broadcast::Sender<Arc<MessagingEvent>>,
+    ) -> Self
+    {
+        Self {
+            peer,
+            inbound_message_tx,
+            messaging_events_tx,
+        }
+    }
+
+    pub async fn run<S>(mut self, socket: S)
+    where S: AsyncRead + AsyncWrite + Unpin {
+        let mut framed_socket = MessagingProtocol::framed(socket);
+        let peer = &self.peer;
+        while let Some(result) = framed_socket.next().await {
+            match result {
+                Ok(raw_msg) => {
+                    trace!(
+                        target: LOG_TARGET,
+                        "Received message from peer '{}' ({} bytes)",
+                        peer.node_id.short_str(),
+                        raw_msg.len()
+                    );
+
+                    let inbound_msg = InboundMessage::new(Arc::clone(&peer), raw_msg.freeze());
+
+                    let event = MessagingEvent::MessageReceived(
+                        Box::new(inbound_msg.source_peer.node_id.clone()),
+                        inbound_msg.tag,
+                    );
+
+                    if let Err(err) = self.inbound_message_tx.send(inbound_msg).await {
+                        warn!(
+                            target: LOG_TARGET,
+                            "Failed to send InboundMessage for peer '{}' because '{}'",
+                            peer.node_id.short_str(),
+                            err
+                        );
+
+                        if err.is_disconnected() {
+                            break;
+                        }
+                    }
+
+                    trace!(target: LOG_TARGET, "Inbound handler sending event '{}'", event);
+                    if let Err(err) = self.messaging_events_tx.send(Arc::new(event)) {
+                        trace!(
+                            target: LOG_TARGET,
+                            "Messaging event '{}' not sent for peer '{}' because there are no subscribers. \
+                             MessagingEvent dropped",
+                            err.0,
+                            peer.node_id.short_str(),
+                        );
+                    }
+                },
+                Err(err) => {
+                    error!(
+                        target: LOG_TARGET,
+                        "Failed to receive from peer '{}' because '{}'",
```

Comment:
I prefer using debug for displaying errors as display only does to_string which hides the under laying error if there is one.

Reply:
This is a `std::io::Error` which has a good Display impl